mantidproject/mantid
# Add algorithm for saving ISIS reduced Reflectometry data into the ORSO ASCII file format

### Description of work

#### Summary of work

This PR adds the `SaveISISReflectometryORSO` algorithm to allow saving of reduced Reflectometry data from experiments at ISIS into the ASCII version of the ORSO file format. This is a first, basic version with further additions to come. This PR also integrates the algorithm into the ISIS Reflectometry interface to allow it to be called as part of reductions performed via the GUI.



Fixes #31073. 
Fixes #36408. 




#### Further detail of work

This PR includes the following changes:

- Adds the orsopy library as a dependency to Mantid. This library can be populated with the data and metadata you want to include in the file and then automatically handles formatting and writing the file. This library is being developed specifically to support the ORSO standard and is expected to evolve as the specification evolves, hopefully making it easier for us to keep up to date with changes.
- Adds some `orso_helper` classes to wrap orsopy functionality that might be useful for other facilities if they want to implement their own ORSO save algorithms in Mantid.
- Adds the `SaveISISReflectometryORSO` algorithm for finding relevant information specifically from an ISIS reduced Reflectometry workspace to include in the ORSO file.
- Adds the ORSO ASCII file format as an option to the Save ASCII tab of the ISIS Reflectometry interface so that it can be used as part of GUI reductions. This required some refactoring of the code in this area to support multiple save algorithms. At the moment, there is only an ASCII version of the ORSO file, however a Nexus version of the file is expected in future (support for this is expected to be added to the orsopy library as an additional save method). We will need to consider this at the time, but I expect the Save ASCII tab in the Reflectometry GUI will eventually need to become just a Save tab and support non-ASCII file formats too.

### To test:


1) Go to `Interfaces -> Reflectometry -> ISIS Reflectometry`.
1) On the Runs tab, enter Investigation ID `1120015` and Cycle `11_3`. Click the Search button.
1) From the search results, click to highlight runs `11934`, `11935` and `11936` and click the Transfer button to transfer them into the batch table on the right hand side.
1) Go to the Save ASCII tab. Set a Save Path for saving the file to.
1) In the File Format section, select `ORSO ASCII (*.ort)` from the drop-down. Check that the header and separator options are disabled but the Q resolution checkbox stays enabled.
1) Tick the Q resolution checkbox.
1) In the Automatic Save section, tick option `Save as ASCII on completion`.
1) Go back to the Runs tab and click the Process button.
1) The reduction should complete successfully and there should be a successful call to `SaveISISReflectometryORSO` recorded in the messages pane. If you navigate to the save path you set in the Save ASCII tab, you should see a `.ort` file. You should be able to open this in a text editor and see that there is a header populated with some information and 4 columns of data.
1) Go back to the Save ASCII tab. Enter a File Name Prefix in the box.
1) Select one of the `IvsQ` workspaces in the workspace list.
1) Untick the Q resolution checkbox and click the big Save button.
1) Check that the file is saved out with the prefix applied and, when opened, there should only be 3 columns of data (the Q resolution should have been excluded).





---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
